src/processor/operator/copy/copy_node.cpp

```diff
@@ -54,172 +50,164 @@ CopyNode::CopyNode(std::shared_ptr<CopyNodeSharedState> sharedState, CopyNodeInf
+    if (numNodesAppended < localNodeGroup->getNumNodes()) {
+        sharedNodeGroup->appendNodeGroup(localNodeGroup.get(), numNodesAppended);
+    }
 }
 
-std::pair<std::string, common::row_idx_t> CopyNode::getFilePathAndRowIdxInFile() {
-    auto filePath = filePathVector->getValue<ku_string_t>(
-        filePathVector->state->selVector->selectedPositions[0]);
-    auto rowIdxInFile =
-        rowIdxVector->getValue<int64_t>(rowIdxVector->state->selVector->selectedPositions[1]);
-    return {filePath.getAsString(), rowIdxInFile};
+void CopyNode::initGlobalStateInternal(ExecutionContext* context) {
+    if (!isCopyAllowed()) {
+        throw CopyException("COPY commands can only be executed once on a table.");
+    }
+    sharedState->initialize(copyNodeInfo.wal->getDirectory());
 }
 
-void CopyNode::executeInternal(kuzu::processor::ExecutionContext* context) {
-    logCopyWALRecord();
+void CopyNode::executeInternal(ExecutionContext* context) {
+    // CopyNode goes through UNDO log, should be logged and flushed to WAL before making changes.
+    sharedState->logCopyNodeWALRecord(copyNodeInfo.wal);
     while (children[0]->getNextTuple(context)) {
-        std::vector<std::unique_ptr<InMemColumnChunk>> columnChunks;
-        columnChunks.reserve(sharedState->columns.size());
-        auto [startRowIdx, endRowIdx] = getStartAndEndRowIdx(0 /* columnIdx */);
-        auto [filePath, startRowIdxInFile] = getFilePathAndRowIdxInFile();
-        for (auto i = 0u; i < sharedState->columns.size(); i++) {
-            auto columnChunk = sharedState->columns[i]->createInMemColumnChunk(
-                startRowIdx, endRowIdx, &copyNodeInfo.copyDesc);
-            columnChunk->copyArrowArray(
-                *ArrowColumnVector::getArrowColumn(dataColumnVectors[i]), copyStates[i].get());
-            columnChunks.push_back(std::move(columnChunk));
+        auto dataChunkToCopy = resultSet->getDataChunk(0);
+        // All tuples in the resultSet are in the same data chunk.
+        auto numTuplesToAppend = ArrowColumnVector::getArrowColumn(
+            resultSet->getValueVector(copyNodeInfo.dataColumnPoses[0]).get())
+                                     ->length();
+        uint64_t numAppendedTuples = 0;
+        while (numAppendedTuples < numTuplesToAppend) {
+            numAppendedTuples += localNodeGroup->append(
+                resultSet, copyNodeInfo.dataColumnPoses, numTuplesToAppend - numAppendedTuples);
+            if (localNodeGroup->getNumNodes() == StorageConstants::NODE_GROUP_SIZE) {
+                // Current node group is full, flush it and reset it to empty.
+                auto nodeGroupIdx = sharedState->getNextNodeGroupIdx();
+                localNodeGroup->setNodeGroupIdx(nodeGroupIdx);
+                appendNodeGroupToTableAndPopulateIndex(sharedState->table, localNodeGroup.get(),
+                    sharedState->pkIndex.get(), sharedState->pkColumnID);
+            }
+            if (numAppendedTuples < numTuplesToAppend) {
```

Comment:
Passing `startOffset` is doable. But I think slicing is a better idea as it can reduce the passing an extra argument. I've simplified the slicing logic, as now it doesn't involve creating a new data chunk.